### ISSUE TITLE
Not start

### ISSUE THREAD

**Issue by MotorSportF1** [MediaBrowser/Emby#3062]
Update on 3.2.50.0 and doesn't start.

FreeBSD 11 amd64 >> Return to 3.2.40.0 because working good.

Error:
/usr/local/etc/rc.d # /usr/local/bin/mono /usr/local/lib/emby-server/MediaBrowser.Server.Mono.exe -ffmpeg /usr/local/bin/ffmpeg -ffprobe /usr/local/bin/ffprobe -programdata /var/db/emby-server
Info Main: Emby
        Command line: /usr/local/lib/emby-server/MediaBrowser.Server.Mono.exe -ffmpeg /usr/local/bin/ffmpeg -ffprobe /usr/local/bin/ffprobe -programdata /var/db/emby-server
        Operating system: Unix 11.0.0.0
        64-Bit OS: True
        64-Bit Process: True
        User Interactive: False
        Mono: 4.8.1 (4.8.1.0 Tue Sep 12 01:39:18 UTC 2017)
        Processor count: 8
        Program data path: /var/db/emby-server
        Application directory: /usr/local/lib/emby-server
* Assertion at threadpool-ms-io.c:344, condition `fd >= 0' not met

Stacktrace:

=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries
used by your application.
=================================================================

Abort

**Comment by Sm0ke0ut** [comment]
I'm running in the same issue, the stacktrace is identical. Seems to be broken somewhere between `3.2.40.0` and `3.2.40.6`. Let me know if you need additional information.

```shell
/usr/local/lib/emby-server # /usr/local/bin/mono /usr/local/lib/emby-server/MediaBrowser.Server.Mono.exe -ffmpeg /usr/local/bin/ffmpeg -ffprobe /usr/local/bin/ffprobe -programdata /var/db/emby-server
Info Main: Emby
        Command line: /usr/local/lib/emby-server/MediaBrowser.Server.Mono.exe -ffmpeg /usr/local/bin/ffmpeg -ffprobe /usr/local/bin/ffprobe -programdata /var/db/emby-server
        Operating system: Unix 11.0.0.0
        64-Bit OS: True
        64-Bit Process: True
        User Interactive: False
        Mono: 4.8.1 (4.8.1.0 Tue Sep 12 01:39:18 UTC 2017)
        Processor count: 8
        Program data path: /var/db/emby-server
        Application directory: /usr/local/lib/emby-server
* Assertion at threadpool-ms-io.c:344, condition `fd >= 0' not met

Stacktrace:


=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries
used by your application.
=================================================================

Abort
```

**Comment by mabelisle** [comment]
Same here.
Assertion at threadpool-ms-io.c:344, condition `fd >= 0' not met
freebsd 11.0 x64
I had to go back to version 3.2.40

**Comment by LukePulverenti** [comment]
Was any emby log file created under /var/db/emby-server/logs?

**Comment by mabelisle** [comment]
A very short one.

`2017-12-09 22:16:00.474 Info Main: Emby
	Command line: /usr/local/lib/emby-server/MediaBrowser.Server.Mono.exe -ffmpeg /usr/local/bin/ffmpeg -ffprobe /usr/local/bin/ffprobe -programdata /var/db/emby-server
	Operating system: Unix 11.0.0.0
	64-Bit OS: True
	64-Bit Process: True
	User Interactive: False
	Mono: 4.8.1 (4.8.1.0 Tue Sep 12 01:39:18 UTC 2017)
	Processor count: 2
	Program data path: /var/db/emby-server
	Application directory: /usr/local/lib/emby-server`

**Comment by LukePulverenti** [comment]
Is this already released for FreeBSD or did you manually apply the zip file output that we publish?

**Comment by mabelisle** [comment]
I manually applied the zip. The latest package available is still on 3.2.36.0.

**Comment by LukePulverenti** [comment]
[Emby.Mono.zip](https://github.com/MediaBrowser/Emby/files/1545269/Emby.Mono.zip)

Can you try this one as a test? make sure to retain a backup. thanks.

**Comment by LukePulverenti** [comment]
If the issue persists, check if you have any plugin .dll files underneath /var/db/emby-server/plugins. Then try deleting those files. Thanks.

**Comment by mabelisle** [comment]
It seem to do the trick. It doesn't crash anymore

Half the pictures are missing but at least it's running.

**Comment by Sm0ke0ut** [comment]
> Emby.Mono.zip
>
> Can you try this one as a test? make sure to retain a backup. thanks.

This release you're refering to seems to be version `3.2.17.12`, is that correct? I've tried `3.2.50.0` without plugins, but the same error occurs.

**Comment by mabelisle** [comment]
I assume that the version uploaded by Luke was only for troubleshooting.. ?

**Comment by LukePulverenti** [comment]
Correct.

**Comment by woodsb02** [comment]
Hi Luke,
Any update on what you think the issue could be here? I will hold off on releasing the 3.2.50.0 release on FreeBSD for now, until this is resolved.
Regards,
Ben

**Comment by MotorSportF1** [comment]
The plugins are not the problem because fresh install with no data, no plugins and doesn't start.

**Comment by LukePulverenti** [comment]
The results above has helped narrow it down to some startup code that uses the Mono.Posix library to make syscalls to query the OS information. This allows us to differentiate between freebsd and linux. The problem is I have no explanation for the sudden failure because the code hasn't changed, nor have the included libraries.

So I am just going to remove it and have you guys add a command line param to specify freebsd customizations from the outside in. Might as well do it now because the library will not be usable anyway down the road with .net core.

**Comment by LukePulverenti** [comment]
@woodsb02 

For next update, please add the following to the startup command line.

-os freebsd

Thanks.

**Comment by LukePulverenti** [comment]
In the meantime if you could test the latest beta zip, that would be helpful:
https://github.com/MediaBrowser/Emby/releases/download/3.2.50.5/Emby.Mono.zip

Thanks.

**Comment by woodsb02** [comment]
By using mono --trace, the output surrounding the error is:

```
[0x8009ed000: 6.35947 13] ENTER: System.IOSelectorJob:.ctor (System.IOOperation,System.IOAsyncCallback,System.IOAsyncResult)(this:0x8020b5710[System.IOSelectorJob MediaBrowser.Server.Mono.exe], 1, [System.IOAsyncCallback:0x8020a39a0], [System.Net.Sockets.SocketAsyncResult:0x8020b55f8], )
[0x8009ed000: 6.35948 13] LEAVE: System.IOSelectorJob:.ctor (System.IOOperation,System.IOAsyncCallback,System.IOAsyncResult)
[0x8009ed000: 6.35968 13] ENTER: System.Net.Sockets.Socket:QueueIOSelectorJob (System.Collections.Generic.Queue`1<System.Collections.Generic.KeyValuePair`2<intptr, System.IOSelectorJob>>,intptr,System.IOSelectorJob)(this:0x8020b5418[System.Net.Sockets.Socket MediaBrowser.Server.Mono.exe], XX, 0xffffffff, [System.IOSelectorJob:0x8020b5710], )
[0x8009ed000: 6.35984 14] ENTER: System.Collections.Generic.Queue`1<System.Collections.Generic.KeyValuePair`2<intptr, System.IOSelectorJob>>:Enqueue (System.Collections.Generic.KeyValuePair`2<intptr, System.IOSelectorJob>)(this:0x8020b5468[System.Collections.Generic.Queue`1 MediaBrowser.Server.Mono.exe], XX, )
[0x8009ed000: 6.35986 14] LEAVE: System.Collections.Generic.Queue`1<System.Collections.Generic.KeyValuePair`2<intptr, System.IOSelectorJob>>:Enqueue (System.Collections.Generic.KeyValuePair`2<intptr, System.IOSelectorJob>)
[0x8009ed000: 6.35995 14] ENTER: (wrapper managed-to-native) System.IOSelector:Add (intptr,System.IOSelectorJob)(0xffffffff, [System.IOSelectorJob:0x8020b5710], )
[0x8009ed000: 6.36011 14] LEAVE: (wrapper managed-to-native) System.IOSelector:Add (intptr,System.IOSelectorJob)
* Assertion at threadpool-ms-io.c:344, condition `fd >= 0' not met

[0x8009ed000: 6.36013 13] LEAVE: System.Net.Sockets.Socket:QueueIOSelectorJob (System.Collections.Generic.Queue`1<System.Collections.Generic.KeyValuePair`2<intptr, System.IOSelectorJob>>,intptr,System.IOSelectorJob)
[0x8009ed000: 6.36016 12] LEAVE: System.Net.Sockets.Socket:ReceiveAsync (System.Net.Sockets.SocketAsyncEventArgs)result=1
Stacktrace:

[0x8009ed000: 6.36017 11] LEAVE: System.Net.NetworkInformation.LinuxNetworkChange:EnsureSocket ()result=1

=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries
used by your application.
=================================================================

[0x8009ed000: 6.36019 11] ENTER: System.Delegate:Combine (System.Delegate,System.Delegate)(0x0, [System.Net.NetworkInformation.NetworkAddressChangedEventHandler:0x8020a30e0], )
```

Comparing the difference between Emby 3.2.40.0 and 3.2.50.0 with the following URL:
https://github.com/MediaBrowser/Emby/compare/3.2.40.0...3.2.50.0

Looking for changes which relate to the trace output above, I am suspicious of the following commit... could it have perhaps introduced this error?
https://github.com/MediaBrowser/Emby/commit/03923a53e014f82a60a72ab0bf4a4d6527517c45

**Comment by LukePulverenti** [comment]
it's possible  yes. i think the mono runtime has a defect with those events if i recall correctly. i guess i could exclude them from freebsd.

**Comment by mabelisle** [comment]
I confirm @woodsb02 theory. Commenting line 27 to 58 of Emby.Server.Implementations/Networking/NetworkManager.cs resolve the issue.

**Comment by LukePulverenti** [comment]
This is resolved in the latest beta zips, but as mentioned, please add the following to the startup command line:

-os "freebsd"

Thanks.

**Comment by mabelisle** [comment]
I confirm the issue is fixed on 3.2.50.11-beta. I added `-os freebsd` like @LukePulverenti instructed at the end of the command_args in `/usr/local/etc/rc.d/emby-server`.

```
command_args="-f -p ${pidfile} ${procname} /usr/local/lib/emby-server/MediaBrowser.Server.Mono.exe -ffmpeg /usr/local/bin/ffmpeg -ffprobe /usr/l
ocal/bin/ffprobe -programdata ${emby_server_data_dir} -os freebsd"
```

**I had to use vi to edit the file because nano insert a carriage return in the command_args line and make the script unusable**

I count on @woodsb02 to update the port ;) Thanks again guys! Good job!
